internal/clients/http/gcp/gcp_service_client.go

```diff
@@ -0,0 +1,175 @@
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to list regions and zones: %w", err)
+	}
+	defer client.Close()
+
+	// This request returns approx. 6kB response (gzipped). Although REST API allows allow-list
+	// of fields via the 'fields' URL param ("items.name,items.zones"), gRPC does not allow that.
+	// Therefore, we must download all the information only to extract region and zone names.
+	req := &computepb.ListRegionsRequest{
+		Project: c.auth.Payload,
+	}
+	iter := client.List(ctx, req)
+	regions := make([]clients.Region, 0, 32)
+	zones := make([]clients.Zone, 0, 64)
+	for {
+		region, err := iter.Next()
+		if errors.Is(err, iterator.Done) {
+			break
+		}
+		if err != nil {
+			return nil, nil, fmt.Errorf("iterator error: %w", err)
+		}
+		regions = append(regions, clients.Region(*region.Name))
+		for _, zone := range region.Zones {
+			zoneAsArr := strings.Split(zone, "/")
+			zones = append(zones, clients.Zone(zoneAsArr[len(zoneAsArr)-1]))
+		}
+	}
+	return regions, zones, nil
+}
+
+func (c *gcpServiceClient) newMachineTypeClient(ctx context.Context) (*compute.MachineTypesClient, error) {
+	client, err := compute.NewMachineTypesRESTClient(ctx, c.options...)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create GCP regions client: %w", err)
+	}
+	return client, nil
+}
+
+func (c *gcpServiceClient) ListMachineTypes(ctx context.Context, zone string) ([]*clients.InstanceType, error) {
+	client, err := c.newMachineTypeClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("unable to list machine types: %w", err)
+	}
+	defer client.Close()
+
+	if zone == "" {
+		zone = config.GCP.DefaultZone
+	}
+
+	req := &computepb.ListMachineTypesRequest{
+		Project: c.auth.Payload,
+		Zone:    zone,
+	}
+	iter := client.List(ctx, req)
+	machineTypes := make([]*clients.InstanceType, 0, 32)
+	for {
+		machineType, err := iter.Next()
+		if errors.Is(err, iterator.Done) {
+			break
+		}
+		if err != nil {
+			return nil, fmt.Errorf("iterator error: %w", err)
+		}
+		arch := clients.ArchitectureTypeX86_64
+		if getMachineFamily(*machineType.Name) == "t2a" {
+			arch = clients.ArchitectureTypeArm64
+		}
+		machineTypes = append(machineTypes, &clients.InstanceType{
+			Name:               clients.InstanceTypeName(*machineType.Name),
+			VCPUs:              *machineType.GuestCpus,
+			Cores:              0,
+			Architecture:       arch,
+			MemoryMiB:          mbToMib(float32(*machineType.MemoryMb)),
+			EphemeralStorageGB: getTotalStorage(machineType.ScratchDisks),
+		})
+	}
+	return machineTypes, nil
+}
+
+func getZone(zone clients.Zone) string {
+	zoneAsArr := strings.Split(zone.String(), "/")
+	return zoneAsArr[len(zoneAsArr)-1]
+}
+
+func getMachineFamily(machineType string) string {
+	machineFamily := strings.Split(machineType, "-")
+	return machineFamily[0]
+}
+
+func getTotalStorage(disks []*computepb.ScratchDisks) int64 {
+	var sum int64 = 0
+	for _, disk := range disks {
+		sum = sum + int64(*disk.DiskGb)
```

Comment:
```suggestion
		sum += int64(*disk.DiskGb)
```
I think this should work as well and looks cleaner.

Reply:
Sure let's do that.